wgpu-hal/src/metal/surface.rs

```diff
@@ -74,88 +74,54 @@ impl super::Surface {
-            // If the main layer is not a CAMetalLayer, we create a CAMetalLayer sublayer and use it instead.
-            // Unlike on macOS, we cannot replace the main view as UIView does not allow it (when NSView does).
+            // If the main layer is not a CAMetalLayer, we create a CAMetalLayer and use it.
             let new_layer: mtl::MetalLayer = msg_send![class, new];
-            let bounds: CGRect = msg_send![main_layer, bounds];
-            let () = msg_send![new_layer.as_ref(), setFrame: bounds];
-            let () = msg_send![main_layer, addSublayer: new_layer.as_ref()];
-            new_layer
-        };
-
-        let window: *mut Object = msg_send![view, window];
-        if !window.is_null() {
-            let screen: *mut Object = msg_send![window, screen];
-            assert!(!screen.is_null(), "window is not attached to a screen");
-
-            let scale_factor: CGFloat = msg_send![screen, nativeScale];
-            let () = msg_send![view, setContentScaleFactor: scale_factor];
-        }
-
-        let _: *mut c_void = msg_send![view, retain];
-        Self::new(NonNull::new(view), render_layer)
-    }
-
-    #[cfg(target_os = "macos")]
-    #[allow(clippy::transmute_ptr_to_ref)]
-    pub unsafe fn from_nsview(
-        nsview: *mut c_void,
-        delegate: &HalManagedMetalLayerDelegate,
-    ) -> Self {
-        let view = nsview as *mut Object;
-        if view.is_null() {
-            panic!("window does not have a valid contentView");
-        }
-
-        let class = class!(CAMetalLayer);
-        // Deprecated! Clients should use `create_surface_from_layer` instead.
-        let is_actually_layer: BOOL = msg_send![view, isKindOfClass: class];
-        if is_actually_layer == YES {
-            return Self::from_layer(mem::transmute(view));
-        }
-
-        let existing: *mut Object = msg_send![view, layer];
-        let use_current = if existing.is_null() {
-            false
-        } else {
-            let result: BOOL = msg_send![existing, isKindOfClass: class];
-            result == YES
-        };
-
-        let render_layer: mtl::MetalLayer = if use_current {
-            mem::transmute::<_, &mtl::MetalLayerRef>(existing).to_owned()
-        } else {
-            let layer: mtl::MetalLayer = msg_send![class, new];
-            let () = msg_send![view, setLayer: layer.as_ref()];
-            let () = msg_send![view, setWantsLayer: YES];
-            let bounds: CGRect = msg_send![view, bounds];
-            let () = msg_send![layer.as_ref(), setBounds: bounds];
-
-            let window: *mut Object = msg_send![view, window];
-            if !window.is_null() {
-                let scale_factor: CGFloat = msg_send![window, backingScaleFactor];
-                let () = msg_send![layer, setContentsScale: scale_factor];
+            let frame: CGRect = msg_send![main_layer, bounds];
+            let () = msg_send![new_layer.as_ref(), setFrame: frame];
+            #[cfg(target_os = "ios")]
```

Comment:
can we use more type checked variant, i.e. `if cfg!(target_os = ...) {...} else if ... {...}`

Reply:
`cfg!` is a runtime condition and `kCAGravityTopLeft` only available on macOS target, compiling to iOS target cause error:
```sh
error[E0425]: cannot find value `kCAGravityTopLeft` in this scope
   --> /Users/LiJinlei/Rust/forks/wgpu/wgpu-hal/src/metal/surface.rs:109:76
    |
109 |                 let () = msg_send![new_layer.as_ref(), setContentsGravity: kCAGravityTopLeft];
    |                                                                            ^^^^^^^^^^^^^^^^^ not found in this scope

For more information about this error, try `rustc --explain E0425`.
```